src/ccl_power.c

```diff
@@ -406,194 +443,123 @@ static void ccl_cosmology_compute_power_class(ccl_cosmology * cosmo, int * statu
   if (*status != CCL_ERROR_CLASS)
     ccl_run_class(cosmo, &fc,&pr,&ba,&th,&pt,&tr,&pm,&sp,&nl,&le,&op,init_arr,status);
 
-  if (*status == CCL_ERROR_CLASS) {
-    //printed error message while running CLASS
-    ccl_free_class_structs(cosmo, &ba,&th,&pt,&tr,&pm,&sp,&nl,&le,init_arr,status);
-    
-    // free the parser
-    parser_free(&fc);
-    return;
-  }
-
   if (parser_free(&fc)== _FAILURE_) {
     *status = CCL_ERROR_CLASS;
     ccl_cosmology_set_status_message(cosmo, "ccl_power.c: ccl_cosmology_compute_power_class(): Error freeing CLASS parser\n");
-    ccl_free_class_structs(cosmo, &ba,&th,&pt,&tr,&pm,&sp,&nl,&le,init_arr,status);
-    return;
   }
-
-  //These are the limits of the splining range
-  cosmo->data.k_min_lin=2*exp(sp.ln_k[0]);
-  cosmo->data.k_max_lin=ccl_splines->K_MAX_SPLINE;
-
-  //CLASS calculations done - now allocate CCL splines
-  double kmin = cosmo->data.k_min_lin;
-  double kmax = ccl_splines->K_MAX_SPLINE;
-  //Compute nk from number of decades and N_K = # k per decade
-  double ndecades = log10(kmax) - log10(kmin);
-  int nk = (int)ceil(ndecades*ccl_splines->N_K);
-  double amin = ccl_splines->A_SPLINE_MINLOG_PK;
-  double amax = ccl_splines->A_SPLINE_MAX;
-  int na = ccl_splines->A_SPLINE_NA_PK+ccl_splines->A_SPLINE_NLOG_PK-1;
-
-  // The x array is initially k, but will later
-  // be overwritten with log(k)
-  double * x = ccl_log_spacing(kmin, kmax, nk);
-  double * a = ccl_linlog_spacing(amin, ccl_splines->A_SPLINE_MIN_PK, amax, ccl_splines->A_SPLINE_NLOG_PK, ccl_splines->A_SPLINE_NA_PK);
-  double * y2d_lin = malloc(nk * na * sizeof(double));
-  double * y2d_nl = malloc(nk * na * sizeof(double));
-
   
-  //If error, store status, we will free later
-  if (a==NULL|| x==NULL || y2d_lin==NULL || y2d_nl==NULL) {
-    *status = CCL_ERROR_SPLINE;
-    ccl_cosmology_set_status_message(cosmo, "ccl_power.c: ccl_cosmology_compute_power_class(): memory allocation error\n");
+  double kmin,kmax,ndecades,amin,amax,ic;
+  int nk,na,s;
+  double *x=NULL, *z=NULL, *y2d_lin=NULL, *y2d_nl=NULL;
+  if (*status == 0) {
+    //CLASS calculations done - now allocate CCL splines
+    kmin = 2*exp(sp.ln_k[0]);
```

Comment:
Why the factor of 2 here?

Reply:
Or should the min be the CCL value?

Reply:
I'd point you to Elisabeth or Elisa (they implemented the CLASS interface, and this particular line hasn't been changed in this PR). I imagine this is to make sure that we don't try to evaluate things below the minimum k used by CLASS.

Reply:
@elisachisari @elikrause can you answer the question here?